### ISSUE TITLE
Calico + rkt Fails due to CNI error

### ISSUE THREAD

**Issue by ElijahCaine** [coreos/coreos-kubernetes#801]
This was discussed in the `#kubernetes-users` channel on the K8s Slack but I wanted to make sure it was documented here.

## TLDR

When Calico + rkt are enabled on the vagrant (and other setups?) multi-node cluster all pods get stuck at the `CreatingContainer` phase because of some problem with CNI not begin setup.

## Problem

Using the latest commit ( 79b7350fe2e45a1a5e9ed0f34a904eb10c158232 ), when `rkt` and `calico` are enabled, ( i.e.,
```bash
# Whether to use Calico for Kubernetes network policy.
export USE_CALICO=true

# Determines the container runtime for kubernetes to use. Accepts 'docker' or 'rkt'.
export CONTAINER_RUNTIME=rkt
```
is set in `controller-install.sh` and `worker-install.sh` ) we get the following error when trying to create a Pod:

```
$ kubectl create -f <spec for deployment with busybox container pod>
$ kubectl get pods
NAME                         READY     STATUS              RESTARTS   AGE
app-deploy-172403538-gvwpz   0/1       ContainerCreating   0          1m
$ kubectl describe pod app-deploy-172403538-gvwpz
[...]
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  44s		29s		6	{default-scheduler }			Warning		FailedScheduling	no nodes available to schedule pods
  12s		12s		1	{default-scheduler }			Normal		Scheduled		Successfully assigned app-deploy-172403538-gvwpz to 172.17.4.202
  12s		1s		2	{kubelet 172.17.4.202}			Warning		FailedSync		Error syncing pod, skipping: failed to SyncPod: failed to set up pod network: cni config unintialized
```

Nodes have possibly have a similar story:

```
kubectl describe node <node name>
[...]
 OS Image:			Container Linux by CoreOS 1284.0.0 (Ladybug)
[...]
 Container Runtime Version:	rkt://1.21.0
 Kubelet Version:		v1.5.1+coreos.0
 Kube-Proxy Version:		v1.5.1+coreos.0
[...]
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  5m		5m		1	{kubelet 172.17.4.101}			Warning		ImageGCFailed		unable to find data for container /
[... everything else normal ...]
```

## Specs

```
$ vagrant version
Installed Version: 1.8.6
Latest Version: 1.9.1
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-14T00:57:05Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1+coreos.0", GitCommit:"cc65f5321f9230bf9a3fa171155c1213d6e3480e", GitTreeState:"clean", BuildDate:"2016-12-14T04:08:28Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

## Hints

It was mentioned in channel that this is because of the Hyperkube image and it's compatibility with rkt [and/or] calico. I wasn't sure how to put that theory to the test so I can't confirm it.

**Comment by yifan-gu** [comment]
For some reason, the `/etc/kubernetes/cni/net.d/` is empty when `USE_CALICO=true`. Not sure whether that's expected.

**Comment by yifan-gu** [comment]
cc @heschlie

**Comment by heschlie** [comment]
Taking a look at this, rkt seems to be having some issues creating the Calico containers.  I see:

```
Failed to create rkt container with error: json: error calling MarshalJSON for the type *schema.PodManifest: json: error calling MarshalJSON for type types.Volume: source for host volume cannot be empty.
```

The volumes section in the manifest looks ok to me, but without it the container can't come online.  I'm not super familiar with rkt, are there any known issues related to volume mounts?

**Comment by yifan-gu** [comment]
@heschlie Thanks, will try to reproduce by creating the calico pod.

**Comment by Promaethius** [comment]
Having the same error. At first I thought it was because, in kubernetes, rkt errors when mounting to a hostDir that doesn't exist. Whereas, on docker, the daemon creates the directory on the host if it does not exist.

**Comment by kfirufk** [comment]
having the same problem. are you sure that rkt can't mount individual files ?
in https://coreos.com/rkt/docs/latest/subcommands/run.html it says that "host volumes that can expose a directory or a file from the host to the pod." under the title "Mount Volumes into a Pod".

**Comment by Promaethius** [comment]
Rkt has the capability but I don't think that kubernetes is handling single
file mounts correctly. The exception that is being thrown comes from
v.Source = "" which doesn't really make sense.

**Comment by yifan-gu** [comment]
cc @jonboulle @squeed

**Comment by Promaethius** [comment]
It looks like the config map that contains the template for the CNI isn't being passed correctly through rkt. I'm assuming that this is an issue with a multi-line environmental variable since only the first line, which is the {, is being passed. 

My idea for a fix that would work with both docker and rkt is to mount the variable in kubernetes as a file in the container then use an additional command arg in the daemonsetdeployment for calico to cat the file into the environmental variable that the script https://github.com/projectcalico/cni-plugin/blob/master/k8s-install/scripts/install-cni.sh is looking for.

I'll submit a pull request if I'm successful.

**Comment by caseydavenport** [comment]
> There is a new problem however in that calico_node's runtime is docker by default.

Looks like the version of Calico that is shipping in the hyperkube is _really really old_.  Newer versions of Calico do not have that issue, and when using the self-hosted install I would not expect the Calico binaries within the hyperkube container to be used at all.

CC @heschlie

**Comment by caseydavenport** [comment]
> Removing the DNS did end up breaking calico and the CNI files were not fully developed

I do not think Calico should need the /etc/resolv.conf mount.  Can you elaborate on what went wrong when this mount was removed?

**Comment by kfirufk** [comment]
Hi. I tried with the latest versions, still encountered the same problem. https://github.com/projectcalico/cni-plugin/issues/253

**Comment by Promaethius** [comment]
@kfirufk that issue is a rkt issue about accepting a multi-line environmental variable. This was my docker/rkt compatible fix for it:

- name: install-cni
          image: quay.io/calico/cni:v1.5.5
          command: ["/bin/sh", "-c"]
          args: ["export CNI_NETWORK_CONFIG=$(cat /host/cni_network_config/config.conf) && /install-cni.sh"]
          env:
            # The location of the Calico etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_endpoints
            # CNI configuration filename
            - name: CNI_CONF_NAME
              value: "10-calico.conf"
          volumeMounts:
            - mountPath: /host/opt/cni/bin
              name: cni-bin-dir
            - mountPath: /host/etc/cni/net.d
              name: cni-net-dir
            - mountPath: /calico-secrets
              name: etcd-certs
            # The CNI network config to install on each node.
            - mountPath: /host/cni_network_config
              name: cni-config

With the configmap mounted as a volume in:
volumes:
        # Used by calico/node.
        - name: lib-modules
          hostPath:
            path: /lib/modules
        - name: var-run-calico
          hostPath:
            path: /var/run/calico
        # Used to install CNI.
        - name: cni-bin-dir
          hostPath:
            path: /opt/cni/bin
        - name: cni-net-dir
          hostPath:
            path: /etc/kubernetes/cni/net.d
        # Mount in the etcd TLS secrets.
        - name: etcd-certs
          secret:
            secretName: calico-etcd-secrets
        - name: cni-config
          configMap:
            name: calico-config
            items:
            - key: cni_network_config
              path: config.conf

**Comment by Promaethius** [comment]
Btw you have to use cat <<'EOF' so that it echos the correct environment cat command to the file. This means you need to split the node/CNI manifest and the config file into two separate cat sections so the etcd address is echoed correctly to the config file with EOF and the variable is echoed to the CNI/node manifest with 'EOF'.

**Comment by Promaethius** [comment]
@caseydavenport CNI doesn't need it but rkt has a section of global mounts passed into the service file with resolv.conf already specified. When rkt tried to mount it twice, it broke. This is why I removed that volume from the CNI, both because it doesn't need and and because it's already a global mount.

**Comment by Promaethius** [comment]
@caseydavenport Also included in those global mounts are the calico binaries. I'm still not understanding why hyperkube is using the old binaries when the new ones are mounted globally I to all the hyperkube rkt instances.

**Comment by Promaethius** [comment]
Turns out most of the issues surrounding the calico binaries mounting to rkt was the if statement:
`# To run a self hosted Calico install it needs to be able to write to the CNI dir
if [ "${USE_CALICO}" = "true" ]; then
    export CALICO_OPTS="--volume cni-bin,kind=host,source=/opt/cni/bin \
                        --mount volume=cni-bin,target=/opt/cni/bin"
else
    export CALICO_OPTS=""
fi
`
Which I placed into the init_function before the templates in the form:
`    # To run a self hosted Calico install it needs to be able to write to the CNI dir
    if [ ${USE_CALICO} = "true" ]; then
        local CALICO_OPTS="--volume cni-bin,kind=host,source=/opt/cni/bin \
        --mount volume=cni-bin,target=/opt/cni/bin"
        echo "RKT Configured for Calico Binaries"
    else
        local CALICO_OPTS=""
    fi
`

The correct binaries now mount, however, there is a known issue preventing from felix running on the node daemons:

https://github.com/kubernetes/minikube/issues/726

**Comment by kfirufk** [comment]
@Promaethius, thanks for your response. I tried to apply your fix on canal.yaml and I get the same results, I probably miss something.

this is my canal.yaml, please let me know what I'm missing :)
```

# This ConfigMap can be used to configure a self-hosted Canal installation.
kind: ConfigMap
apiVersion: v1
metadata:
  name: canal-config
  namespace: kube-system
data:
  # Configure this with the location of your etcd cluster.
  etcd_endpoints: "http://127.0.0.1:4001"

  # The interface used by canal for host <-> host communication.
  # If left blank, then the interface is chosing using the node's
  # default route.
  canal_iface: ""

  # Whether or not to masquerade traffic to destinations not within
  # the pod network.
  masquerade: "true"

  # The CNI network configuration to install on each node.  The special
  # values in this config will be automatically populated.
  cni_network_config: |-
    {
        "name": "canal",
        "type": "flannel",
        "delegate": {
          "type": "calico",
          "etcd_endpoints": "__ETCD_ENDPOINTS__",
          "log_level": "info",
          "policy": {
              "type": "k8s",
              "k8s_api_root": "https://__KUBERNETES_SERVICE_HOST__:__KUBERNETES_SERVICE_PORT__",
              "k8s_auth_token": "__SERVICEACCOUNT_TOKEN__"
          },
          "kubernetes": {
              "kubeconfig": "/etc/cni/net.d/__KUBECONFIG_FILENAME__"
          }
        }
    }

---

# This manifest installs the per-node agents, as well
# as the CNI plugins and network config on
# each master and worker node in a Kubernetes cluster.
kind: DaemonSet
apiVersion: extensions/v1beta1
metadata:
  name: canal-node
  namespace: kube-system
  labels:
    k8s-app: canal-node
spec:
  selector:
    matchLabels:
      k8s-app: canal-node
  template:
    metadata:
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
        scheduler.alpha.kubernetes.io/tolerations: '[{"key":"CriticalAddonsOnly", "operator":"Exists"}]'
      labels:
        k8s-app: canal-node
    spec:
      hostNetwork: true
      containers:
        # Runs the flannel daemon to enable vxlan networking between
        # container hosts.
        - name: flannel
          image: quay.io/coreos/flannel:v0.7.0
          env:
            # The location of the etcd cluster.
            - name: FLANNELD_ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: etcd_endpoints
            # The interface flannel should run on.
            - name: FLANNELD_IFACE
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: canal_iface
            # Perform masquerade on traffic leaving the pod cidr.
            - name: FLANNELD_IP_MASQ
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: masquerade
            # Write the subnet.env file to the mounted directory.
            - name: FLANNELD_SUBNET_FILE
              value: "/run/flannel/subnet.env"
          securityContext:
            privileged: true
          volumeMounts:
            - mountPath: /etc/resolv.conf
              name: resolv
            - mountPath: /run/flannel
              name: run-flannel
        # Runs calico/node container on each Kubernetes node.  This
        # container programs network policy and local routes on each
        # host.
        - name: calico-node
          image: quay.io/calico/node:v1.0.0-13-g313c7ca
          env:
            # The location of the etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: etcd_endpoints
            # Disable Calico BGP.  Calico is simply enforcing policy.
            - name: CALICO_NETWORKING
              value: "false"
            # Disable file logging so `kubectl logs` works.
            - name: CALICO_DISABLE_FILE_LOGGING
              value: "true"
          securityContext:
            privileged: true
          volumeMounts:
            - mountPath: /lib/modules
              name: lib-modules
              readOnly: true
            - mountPath: /var/run/calico
              name: var-run-calico
              readOnly: false
        # This container installs the Calico CNI binaries
        # and CNI network config file on each node.
        - name: install-calico-cni
          image: calico/cni:v1.5.5-10-ge2792bb
          imagePullPolicy: Always
          command:  ["/bin/sh", "-c"]
          args: ["export CNI_NETWORK_CONFIG=$(cat /host/cni_network_config/config.conf) && /install-cni.sh"]
          env:
            # The name of the CNI network config file to install.
            - name: CNI_CONF_NAME
              value: "10-canal.conf"
            # The location of the etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: etcd_endpoints
            # The CNI network config to install on each node.
            - name: CNI_NETWORK_CONFIG
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: cni_network_config
          volumeMounts:
            - mountPath: /host/opt/cni/bin
              name: cni-bin-dir
            - mountPath: /host/etc/cni/net.d
              name: cni-net-dir
            - mountPath: /host/cni_network_config
              name: cni-config
      volumes:
        # Used by calico/node.
        - name: lib-modules
          hostPath:
            path: /lib/modules
        - name: var-run-calico
          hostPath:
            path: /var/run/calico
        # Used to install CNI.
        - name: cni-bin-dir
          hostPath:
            path: /opt/cni/bin
        - name: cni-net-dir
          hostPath:
            path: /etc/cni/net.d
        # Used by flannel daemon.
        - name: run-flannel
          hostPath:
            path: /run/flannel
        - name: resolv
          hostPath:
            path: /etc/resolv.conf
        - name: cni-config
          configMap:
           name: canal-config
           items:
            - key: cni_network_config
              path: config.conf

---

# This manifest deploys a Job which performs one time
# configuration of Canal.
apiVersion: batch/v1
kind: Job
metadata:
  name: configure-canal
  namespace: kube-system
  labels:
    k8s-app: canal
spec:
  template:
    metadata:
      name: configure-canal
    spec:
      hostNetwork: true
      restartPolicy: OnFailure
      containers:
        # Writes basic flannel configuration to etcd.
        - name: configure-flannel
          image: quay.io/coreos/etcd:v3.1.0
          command:
          - "etcdctl"
          - "--no-sync"
          - "set"
          - "/coreos.com/network/config"
          - '{ "Network": "192.168.0.0/16", "Backend": {"Type": "vxlan"} }'
          env:
            # The location of the etcd cluster.
            - name: ETCDCTL_PEERS
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: etcd_endpoints

---

# This manifest deploys the Calico policy controller on Kubernetes.
# See https://github.com/projectcalico/k8s-policy
apiVersion: extensions/v1beta1
kind: ReplicaSet
metadata:
  name: calico-policy-controller
  namespace: kube-system
  labels:
    k8s-app: calico-policy
spec:
  # The policy controller can only have a single active instance.
  replicas: 1
  template:
    metadata:
      name: calico-policy-controller
      namespace: kube-system
      labels:
        k8s-app: calico-policy
    spec:
      # The policy controller must run in the host network namespace so that
      # it isn't governed by policy that would prevent it from working.
      hostNetwork: true
      containers:
        - name: calico-policy-controller
          image: calico/kube-policy-controller:v0.5.2-5-gfa61293
          env:
            # The location of the Calico etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: canal-config
                  key: etcd_endpoints
            # The location of the Kubernetes API.  Use the default Kubernetes
            # service for API access.
            - name: K8S_API
              value: "https://coreos-2.tux-in.com:443"
            # Since we're running in the host namespace and might not have KubeDNS
            # access, configure the container's /etc/hosts to resolve
            # kubernetes.default to the correct service clusterIP.
            - name: CONFIGURE_ETC_HOSTS
              value: "true"
```

**Comment by Promaethius** [comment]
@kfirufk it turns out rkt is intrinsically flawed when running privileged containers. The node daemon felix requires RW privilages to /proc/sys which is mounted incorrectly. I'm trying to open dialogue to find a work around otherwise the bugfix is posted for rkt milestone 1.24. I've referenced the issue in my last post and will post back when I hear something

**Comment by Promaethius** [comment]
Going to try something when I get home tonight and run the calico node pod with the k8s annotation: `rkt.alpha.kubernetes.io/stage1-name-override: coreos.com/rkt/stage1-fly`
This is supposed to give the node the chroot isolation and RW permissions it needs to access the /proc/sys sockets with full privilages.

**Comment by Promaethius** [comment]
I got rktnetes + calico working flawlessly with a cross compatible script. You can do a comparison to my fork of this GitHub project to see what I did because the explanation is too much for my mobile screen to handle right now.

**Comment by kfirufk** [comment]
@Promaethius - yay! great news! i'll look into it this weekend.

**Comment by leodotcloud** [comment]
@Promaethius Could you please share your solution? I couldn't find any commits on your profile page related to this.

**Comment by kfirufk** [comment]
@Promaethius - there's something I don't understand in multi-node/generic/controller-install.sh script of your forked project. the start_addons function uses "/host/manifests" directory and I don't run it from a container but from the ContainerOS itself, so.. the manifests are actually located under "/srv/kubernetes/manifests". I'm probably missing something, but what? :)

**Comment by Promaethius** [comment]
@kfirufk sorry about that I made some changes today and had the wrong directory to serve the addons. I'm making some changes to include cephfs as the storage provider by default so kubernetes is a little more flexible by default. As far as calico goes, however, did you have any questions?

**Comment by kfirufk** [comment]
@Promaethius, I didn't finish applying all your changes. working on forking your script in a easier editable method by actually having each file in a relevant directory instead of one big bash script. regarding calico i'll know soon enough :) I can't seem to find how to open bug reports in your forked project. 40-ExecStartPre-symlink.conf.conf should be one ".conf" i'm guessing :)

**Comment by Promaethius** [comment]
@kfirufk by all means edit it. Feel free to submit pull requests by the way. I'll do my best to explain in this thread why I found calico would not work with this generic script:

_CALICO_OPTS_: Using export in the initialization of the script, the CALICO_OPTS variable remained blank when passed to the hyperkube image. The correct CNI binary folder would not be mounted and the newer Calico plugin would not be written resulting in rkt networking failures.
**Original:**
```yaml
# To run a self hosted Calico install it needs to be able to write to the CNI dir
if [ "${USE_CALICO}" = "true" ]; then
    export CALICO_OPTS="--volume cni-bin,kind=host,source=/opt/cni/bin \
                        --mount volume=cni-bin,target=/opt/cni/bin"
else
    export CALICO_OPTS=""
fi
```

**Fixed:**
```yaml
function init_templates {
    local TEMPLATE=/etc/systemd/system/kubelet.service
    local uuid_file="/var/run/kubelet-pod.uuid"
    if [ ! -f $TEMPLATE ]; then
        echo "TEMPLATE: $TEMPLATE"
        mkdir -p $(dirname $TEMPLATE)
    # To run a self hosted Calico install it needs to be able to write to the CNI dir
    if [ ${USE_CALICO} = "true" ]; then
        local CALICO_OPTS="--volume cni-bin,kind=host,source=/opt/cni/bin \
        --mount volume=cni-bin,target=/opt/cni/bin"
		mkdir -p /lib/modules
		mkdir -p /var/run/calico
		mkdir -p /opt/cni/bin
		mkdir -p /etc/kubernetes/cni/net.d
        echo "RKT Configured for Calico Binaries"
    else
        local CALICO_OPTS=""
    fi
...
```

_Calico-Node:_ 

1. Because of a known issue with rkt, images running on anything other than a stage1-fly image would not be able to write into /proc/sys. This would crash the image since calico-node uses an ip socket in that container. Thanks to some folks over at rkt, a temporary fix was suggested.
2. An old version of the node was being used.
3. /etc/resolv.conf would throw errors if mounted with rkt. (This is not needed to function properly since kubernetes already handles nameservice.)

_Calico-CNI:_
Another bug with rkt, it cannot accept multiline environmental variables. For those who had cryptic errors about brackets, reading the written CNI plugin file would immediately notice what went wrong. Only the first line of the configuration was being received which was just a single open {. The docker/rkt compatible fix for this was to mount the conf as a file and then echo it into the correct environmental before running the installer script.

**Original:**
```yaml
kind: DaemonSet
apiVersion: extensions/v1beta1
metadata:
  name: calico-node
  namespace: kube-system
  labels:
    k8s-app: calico-node
spec:
  selector:
    matchLabels:
      k8s-app: calico-node
  template:
    metadata:
      labels:
        k8s-app: calico-node
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
        scheduler.alpha.kubernetes.io/tolerations: |
          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
    spec:
      hostNetwork: true
      containers:
        # Runs calico/node container on each Kubernetes node.  This 
        # container programs network policy and routes on each
        # host.
        - name: calico-node
          image: quay.io/calico/node:v0.23.0
          env:
            # The location of the Calico etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_endpoints
            # Choose the backend to use. 
            - name: CALICO_NETWORKING_BACKEND
              value: "none"
            # Disable file logging so 'kubectl logs' works.
            - name: CALICO_DISABLE_FILE_LOGGING
              value: "true"
            - name: NO_DEFAULT_POOLS
              value: "true"
          securityContext:
            privileged: true
          volumeMounts:
            - mountPath: /lib/modules
              name: lib-modules
              readOnly: false
            - mountPath: /var/run/calico
              name: var-run-calico
              readOnly: false
            - mountPath: /etc/resolv.conf
              name: dns
              readOnly: true
        # This container installs the Calico CNI binaries
        # and CNI network config file on each node.
        - name: install-cni
          image: quay.io/calico/cni:v1.5.2
          imagePullPolicy: Always
          command: ["/install-cni.sh"]
          env:
            # CNI configuration filename
            - name: CNI_CONF_NAME
              value: "10-calico.conf"
            # The location of the Calico etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_endpoints
            # The CNI network config to install on each node.
            - name: CNI_NETWORK_CONFIG
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: cni_network_config
          volumeMounts:
            - mountPath: /host/opt/cni/bin
              name: cni-bin-dir
            - mountPath: /host/etc/cni/net.d
              name: cni-net-dir
      volumes:
        # Used by calico/node.
        - name: lib-modules
          hostPath:
            path: /lib/modules
        - name: var-run-calico
          hostPath:
            path: /var/run/calico
        # Used to install CNI.
        - name: cni-bin-dir
          hostPath:
            path: /opt/cni/bin
        - name: cni-net-dir
          hostPath:
            path: /etc/kubernetes/cni/net.d
        - name: dns
          hostPath:
            path: /etc/resolv.conf
```

**Fixed:**
```yaml
# This manifest installs the calico/node container, as well
# as the Calico CNI plugins and network config on
# each master and worker node in a Kubernetes cluster.
kind: DaemonSet
apiVersion: extensions/v1beta1
metadata:
  name: calico-node
  namespace: kube-system
  labels:
    k8s-app: calico-node
spec:
  selector:
    matchLabels:
      k8s-app: calico-node
  template:
    metadata:
      labels:
        k8s-app: calico-node
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
        scheduler.alpha.kubernetes.io/tolerations: |
          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
    spec:
      hostNetwork: true
      containers:
        # Runs calico/node container on each Kubernetes node.  This
        # container programs network policy and routes on each
        # host.
        - name: calico-node
          image: quay.io/calico/node:v1.0.1
          command: ["/bin/sh", "-c"]
          args: ["mount -o remount,rw /proc/sys && start_runit"]
          env:
            # The location of the Calico etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_endpoints
            # Choose the backend to use.
            - name: CALICO_NETWORKING_BACKEND
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: calico_backend
            # Disable file logging so `kubectl logs` works.
            - name: CALICO_DISABLE_FILE_LOGGING
              value: "true"
            # Don't configure a default pool.  This is done by the Job
            # below.
            - name: NO_DEFAULT_POOLS
              value: "true"
            - name: FELIX_LOGSEVERITYSCREEN
              value: "info"
            # Location of the CA certificate for etcd.
            - name: ETCD_CA_CERT_FILE
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_ca
            # Location of the client key for etcd.
            - name: ETCD_KEY_FILE
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_key
            # Location of the client certificate for etcd.
            - name: ETCD_CERT_FILE
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_cert
            # Auto-detect the BGP IP address.
            - name: IP
              value: ""
          securityContext:
            privileged: true
          volumeMounts:
            - mountPath: /lib/modules
              name: lib-modules
              readOnly: false
            - mountPath: /var/run/calico
              name: var-run-calico
              readOnly: false
            - mountPath: /calico-secrets
              name: etcd-certs
        # This container installs the Calico CNI binaries
        # and CNI network config file on each node.
        - name: install-cni
          image: quay.io/calico/cni:v1.5.5
          command: ["/bin/sh", "-c"]
          args: ["export CNI_NETWORK_CONFIG=$(cat /host/cni_network_config/config.conf) && /install-cni.sh"]
          env:
            # The location of the Calico etcd cluster.
            - name: ETCD_ENDPOINTS
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: etcd_endpoints
            # CNI configuration filename
            - name: CNI_CONF_NAME
              value: "10-calico.conf"
          securityContext:
            privileged: true
          volumeMounts:
            - mountPath: /host/opt/cni/bin
              name: cni-bin-dir
            - mountPath: /host/etc/cni/net.d
              name: cni-net-dir
            - mountPath: /calico-secrets
              name: etcd-certs
            # The CNI network config to install on each node.
            - mountPath: /host/cni_network_config
              name: cni-config
      volumes:
        # Used by calico/node.
        - name: lib-modules
          hostPath:
            path: /lib/modules
        - name: var-run-calico
          hostPath:
            path: /var/run/calico
        # Used to install CNI.
        - name: cni-bin-dir
          hostPath:
            path: /opt/cni/bin
        - name: cni-net-dir
          hostPath:
            path: /etc/kubernetes/cni/net.d
        # Mount in the etcd TLS secrets.
        - name: etcd-certs
          secret:
            secretName: calico-etcd-secrets
        - name: cni-config
          configMap:
            name: calico-config
            items:
            - key: cni_network_config
              path: config.conf
```

**Comment by kfirufk** [comment]
@Promaethius - still having problems starting calico on worker node. when you have a time please take a loot. http://stackoverflow.com/questions/42741272/calico-node-fails-starting-on-worker-node

thanks
